### ISSUE TITLE
IOUring: Add supported for provided buffers

### ISSUE THREAD

**Issue by normanmaurer** [netty/netty#14614]
See https://github.com/axboe/liburing/wiki/What's-new-with-io_uring-in-6.11-and-6.12#incremental-provided-buffer-consumption

**Comment by dreamlike-ocean** [comment]
Can this be implemented in version 4.2? Or do we need to wait for some time?
I encountered several problems when implementing features for my io_uring library before. Maybe it will be helpful to you.

1. whether to support incremental provided buffers or not. Since this feature is quite new, it has a relatively small user base.I didn't implement this feature.
2. there's the problem regarding the buffer size in the buffer ring. I'm not quite sure how much memory should be allocated,So I'll only allocate the memory of one page size.
3. The life cycle of io_uring_buf.  it's possible that users may hold onto this buffer for too long, resulting in the buffer ring having no elements available.So, I'll choose to release it into the buffer ring directly after copying it to the direct memory provided by the users.

**Comment by normanmaurer** [comment]
I don't think we need to have this for 4.2.0.Final but it would be nice to have it at some point. Also I think it should be something that can be enabled and disabled via a `ChannelOption`. I might take a stab on this within the next few weeks

**Comment by dreamlike-ocean** [comment]
Hi, I've completed a prototype. Could we discuss whether this implementation is reasonable or not?

Implement IoUringBufferRing and IOSQE_BUFFER_SELECT flag for socket recv:
- Enable the use of IOSQE_BUFFER_SELECT in the socket recv when IoUringBufferRing is supported by the current Linux kernel version and IOUringSocketChannelConfig is configured to enable the feature
- IoUringBufferRing elements are not populated immediately; only when submitting recv ops, if there is available space in the current BufferRing, the allocator will be used to allocate a buffer.
- If the res in the recv CQE is -ERRNO_NO_BUFFER, mark the bufferRing as temporarily unavailable. The current and subsequent recv operations will fall back to the old recv path.
- When a ByteBuf allocated from IoUringBufferRing is released by the user, it will be returned to the IoUringBufferRing, and the buffer will be marked as available for reuse.

[provide_buffer](https://github.com/dreamlike-ocean/netty/compare/provide_buffer)([#1](https://github.com/dreamlike-ocean/netty/pull/1))

**Comment by normanmaurer** [comment]
This sounds reasonable but I will need to check the code in detail once back from vacation. Do you have numbers to share yet ?

**Comment by dreamlike-ocean** [comment]
> This sounds reasonable but I will need to check the code in detail once back from vacation. Do you have numbers to share yet ?

No performance data yet, I'm just trying to implement it

Have a nice vocation!

**Comment by normanmaurer** [comment]
@dreamlike-ocean please open a PR when you feel there is something ready to review

**Comment by dreamlike-ocean** [comment]
> [@dreamlike-ocean](https://github.com/dreamlike-ocean) please open a PR when you feel there is something ready to review

I've completed the code. Please review this pr.,https://github.com/netty/netty/pull/14690 
please reivew this pr
